src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -103,6 +117,187 @@ SUNAdaptController SUNAdaptController_Soderlind(SUNContext sunctx)
   return (C);
 }
 
+/* -----------------------------------------------------------------
+ * Function to control Soderlind parameters from the command line
+ */
+
+SUNErrCode SUNAdaptController_SetFromCommandLine_Soderlind(SUNAdaptController C,
+                                                           const char* Cid,
+                                                           int argc, char* argv[])
+{
+  SUNFunctionBegin(C->sunctx);
+
+  int i;
+  SUNErrCode retval;
+  sunbooleantype write_parameters = SUNFALSE;
+  for (i = 1; i < argc; i++)
+  {
+    /* if Cid is supplied, skip command-line arguments that do not begin with Cid;
+       else, skip command-line arguments that do not begin with "sunadaptcontroller." */
+    size_t SOffset, PIDOffset, PIOffset, IOffset, ExpGusOffset, ImpGusOffset;
+    const char* SPrefix   = (strlen(Cid) > 0) ? Cid
+                                              : "sunadaptcontroller_soderlind.";
+    const char* PIDPrefix = (strlen(Cid) > 0) ? Cid : "sunadaptcontroller_pid.";
+    const char* PIPrefix  = (strlen(Cid) > 0) ? Cid : "sunadaptcontroller_pi.";
+    const char* IPrefix   = (strlen(Cid) > 0) ? Cid : "sunadaptcontroller_i.";
+    const char* ExpGusPrefix = (strlen(Cid) > 0) ? Cid
+                                                 : "sunadaptcontroller_expgus.";
+    const char* ImpGusPrefix = (strlen(Cid) > 0) ? Cid
+                                                 : "sunadaptcontroller_impgus.";
```

Comment:
I think a couple more controllers were added in 7.3 so this would need to be updated. The bigger issues are this function is difficult to maintain and it breaks what I think is a convention of associating one prefix per class. I'm struggling to think of a good solution, but moving bias, defaults, and write parameter parsing up to superclass will help.

Reply:
I disagree.  The new controllers that were added in 7.3 have fixed coefficients, and thus they do not even provide `SetParams` routines in the first place.  This piece had to be complicated to support our older controller API, where the controller "type" merely indicated the specific structure of the controller (one vs two vs three (h,dsm) history values, etc), but all exponents were user modifiable.

Since the new controllers have no `SetParams` routines, then their other parameters (e.g., bias) can already be specified via the `Cid` value `sunadaptcontroller_soderlind`.  I added this note to the documentation for each of those utility constructors.